File: matlab_files/python_substituted/depth_classifier.m
function [classified_path,files_to_classify,maximum_depth] = depth_classifier(...
    sd_input,deflection_processing_path,individual_path,components_to_plot,...
    complete_files_path,spherical_complete_files_path,headers_on)
% This function takes as input the paths to the complete stress or
% deflection files containing the stress or deflection components and the
% corresponding centroid or node coordinates and discretizes the results
% based on depth. 

% Define parameters that change based on whether we are working with
% stresses or deflections, like plot names and file path parts. 

if sd_input == 0
    upper_path = individual_path;
    save_name = 'Stresses_until_depth_';
    histogram_name_part = 'centroids';
    matrix_to_save_part = 'Stresses_layer_';
    earth_name_part = 'Stresses_Earth_';
    discretized_headers = {'Label','Mises','S_11','S_22','S_33','S_12','S_13',...
        'S_23','X','Y','Z','R','Depth','Lat','Lon'};
else
    upper_path = deflection_processing_path;
    save_name = 'Deflections_until_depth_';
    histogram_name_part = 'nodes';
    matrix_to_save_part = 'Deflections_layer_';
    earth_name_part = 'Deflections_Earth_';
    discretized_headers = {'Label','U_Magn','U_1','U_2','U_3','X','Y','Z',...
        'R','Depth','Lat','Lon'};
end

% Define the necessary path and folders and create them if they are not
% there, differentiating between the two cases: working with cartesian or
% spherical stress components.

unspherical_classified_depth_path = [upper_path '\Depth_classified_files'];
spherical_classified_depth_path = [upper_path '\Spherical_depth_classified_files'];
    
if strcmp(components_to_plot, 'spherical') == 1
    files_to_classify_path = dir([spherical_complete_files_path '\*.csv']);
    files_to_classify = {files_to_classify_path.name};
    classified_path = spherical_classified_depth_path;
    histogram_path = [spherical_classified_depth_path '\histogram_plots'];
    if ~exist(classified_path, 'dir')
        mkdir (classified_path);
    end
    if ~exist(histogram_path, 'dir')
        mkdir (histogram_path);
    end
elseif strcmp(components_to_plot, 'cartesian') == 1
    files_to_classify_path = dir([complete_files_path '\*.csv']);
    files_to_classify = {files_to_classify_path.name};
    classified_path = unspherical_classified_depth_path;
    histogram_path = [unspherical_classified_depth_path '\histogram_plots'];
    if ~exist(classified_path, 'dir')
        mkdir (classified_path)
    end
    if ~exist(histogram_path, 'dir')
        mkdir (histogram_path)
    end
else
    disp('Incorrect input, select either spherical or unspherical.');
end
earth_data_path = [classified_path '\No_layers_subdivision'];
% Only process part files, taking away those relative to the regions.
% Region I1 will be skipped as it only contains empty values while I0 is
% not processed because it belongs to the core, which also does not
% interest us.

for i = 1:length(files_to_classify)
    size_to_evaluate = dir([files_to_classify_path(i).folder '\' files_to_classify{i}]);
    if contains(files_to_classify{i},'Region') == 1
        files_to_classify{i} = {};
    elseif size_to_evaluate.bytes < 100
        files_to_classify{i} = {};
    elseif contains(files_to_classify{i},'I0') == 1
        files_to_classify{i} = {};
    elseif contains(files_to_classify{i},'Large') == 1
        files_to_classify{i} = {};
    end
end
files_to_classify = files_to_classify(~cellfun(@isempty,files_to_classify));
    
% To classify the stress values based on layers, we must build a vector
% containing different layer depths. We know 410 and 660 from the literature, 
% but the depth of  the last one is determined by the maximum element depth 
% ABAQUS has generated, and we must find it.
    
maximum_depth = 0;
earth_radius = 6371000;
earth_data = [];
cfn = length(findobj('type','figure'));

% For every file we read, we need to calculate the radial cemtroid distance
% of every node.

for i = 1:length(files_to_classify)
    % Read the data file and allocate matrices to store depth and radial
    % distance of each centroid or node
    data_matrix = readmatrix([files_to_classify_path(i).folder '\' files_to_classify{i}]);
    layer_depth = zeros(length(data_matrix),1);
    radial_centroid_distance = zeros(length(data_matrix),1);
    % Calculate the radial distance of each centroid or node
    for j = 1:length(data_matrix)
        radial_centroid_distance(j) = sqrt(data_matrix(j,end-2)^2+data_matrix(j,end-1)^2+data_matrix(j,end)^2);
    end
    % Calculate the maximum depth, this will be the lower limit for the
    % lower mantle based on how the model is generated 
    for j = 1:length(radial_centroid_distance)
        radius_to_search = radial_centroid_distance(j);
        layer_depth(j) = earth_radius - radius_to_search;
        if layer_depth(j)>maximum_depth
            maximum_depth = layer_depth(j);
        end
    end
    depthmatrix = [data_matrix,radial_centroid_distance,layer_depth];
    earth_data = [earth_data;depthmatrix];
end

% Differentiate based on whether we study the core too or not

if maximum_depth > 2.886e6
    layer_depth_km = [410,660,2886,maximum_depth/1000];
    layer_depth = layer_depth_km*1000;
else
    layer_depth_km = [410,660,maximum_depth/1000];
    layer_depth = layer_depth_km*1000;
end

check_1 = 1;
check_2 = 1;
run_input = 1;

% Decide whether to run or not the depth classifier proper and if we want
% to change the number of bins we are using

if exist([earth_data_path '\' earth_name_part(1:end-1) '.csv'])
    while check_1 == 1
        run_input = input(['Do you want to rerun the depth discretization algorithm? ' ...
            '1 if yes, 0 if no: \n']);
        if run_input == 0 
            check_1 = 0;
        elseif run_input == 1
            while check_2 == 1
                bin_input = input(['Do you want to change the bin number?' ...
                    ' 1 if yes, 0 if no: \n']);
                if bin_input == 0
                    check_2 = 0;
                    check_1 = 0;
                elseif bin_input == 1
                    n_bins = input('Enter number of layer bins:  \n');
                    earth_bins = input('Enter number of model bins: \n');
                    check_2 = 0;
                    check_1 = 0;
                else
                    disp('incorrect input, select either 1 or 0. \n');
                end
            end
        else
            disp('incorrect input, select either 1 or 0. \n');
        end
    end
end

%--------------------------------------------------------------------------

if run_input == 0
    disp('Skipping this stage to go to geographic plots.')
else
    disp('Starting discretization of data into different depths');
    % We can now create a file for each layer depth based on the values we have
    % just found. First we iterate over the part files we still analyze, then
    % we store the values into the depths given by the vector and we save the
    % files with stresses at depth.

    large_depth_matrix = [];
    
    % Create a complete matrix using the stress or deflection matrix
    % containing stresses or deflections and centroid or nodes coordinates,
    % then appending the centroid distance and depth of each of those
    % points. Also iteratively add matrices to a larger csv file to store
    % the values for each part.
    
    for i = 1:length(files_to_classify)
        data_matrix = readmatrix([files_to_classify_path(i).folder '\' files_to_classify{i}]);
        radial_centroid_distance = zeros(length(data_matrix),1);
        depths = zeros(length(data_matrix),1);

        for k = 1:length(data_matrix)
            radial_centroid_distance(k) = sqrt(data_matrix(k,end-2)^2+data_matrix(k,end-1)^2+data_matrix(k,end)^2);
            depths(k) = earth_radius - radial_centroid_distance(k);
        end

        complete_matrix = [data_matrix, radial_centroid_distance, depths];
        large_depth_matrix = [large_depth_matrix; complete_matrix];
        
    end
    
    if sd_input == 0
        cartesian_coordinates = large_depth_matrix(:,9:11);
    else
        cartesian_coordinates = large_depth_matrix(:,6:8);
    end
    [lat,lon] = cart2geo(cartesian_coordinates);
%     [lat_lon_matrix] = ecef2lla(cartesian_coordinates);
%     lat = lat_lon_matrix(:,1);
%     lon = lat_lon_matrix(:,2);
    large_depth_matrix = [large_depth_matrix,lat,lon];
    
    % After creating the complete large matrix, divide it different layers
    % based on the centroid or nodes depth

    for i = 1:length(layer_depth)
        layer_group_indices = large_depth_matrix(:,end-2) < layer_depth(i);
        layer_matrix = large_depth_matrix(layer_group_indices,:);
        if headers_on == 1
            layer_matrix_wheaders = [discretized_headers;num2cell(layer_matrix)];
            writecell(layer_matrix_wheaders, ...
                [classified_path '\' save_name num2str(layer_depth_km(i)) '_km.csv']);
        else
            writematrix(layer_matrix, ...
                [classified_path '\' save_name num2str(layer_depth_km(i)) '_km.csv']);
        end
        large_depth_matrix(layer_group_indices,:) = 0;
        large_depth_matrix = large_depth_matrix(any(large_depth_matrix,2),:);
        
    end

    % Another step is necessary, and that is now discretizing the data
    % found at each layer in a number of bins decided by the user. After
    % deciding the number of bins, the names of the matrices regarding each
    % layer are read.

    n_bins = 20;
    layer_values = dir([classified_path '\*.csv']);
    layer_values_names = natsortfiles({layer_values.name});

    for i = 1:length(layer_values_names)
        subclassified_path = [classified_path '\subclassification_depth_' ...
            num2str(regexp(layer_values_names{i}, '[\d\.]+', 'match','once')) '_km'];
        if ~exist(subclassified_path, 'dir')
            mkdir (subclassified_path)
        end
        layer_stress_file = readmatrix([classified_path '\' layer_values_names{i}]);
        depth_data = layer_stress_file(:,end-2);
        
        % Generate histograms to show the distribution of centroids or nodes 
        % with depth, based on the number of bins 
        
        figure(cfn+i)
        histogram(depth_data,n_bins);
        xlabel('Depth [m]'); ylabel(['Number of ' histogram_name_part]);
        if i == 1
            title({'Distribution of ' histogram_name_part ' from the ABAQUS model at different depths for',...
                ['layer [0-' num2str(regexp(layer_values_names{i},...
                '[\d\.]+', 'match','once')) '] km']});
            grid on;
            saveas(gcf,[histogram_path '\Layer_[0-' num2str(regexp(layer_values_names{i},...
                '[\d\.]+', 'match','once')) ']_km''_distribution.png'])
        else
            title({'Distribution of ' histogram_name_part ' from the ABAQUS model at different depths for',...
                ['layer [' num2str(regexp(layer_values_names{i-1}, '[\d\.]+', 'match','once'))...
                '-' num2str(regexp(layer_values_names{i}, '[\d\.]+', 'match','once')) '] km']});
            grid on;
            saveas(gcf,[histogram_path '\Layer_[' num2str(regexp(layer_values_names{i-1}, '[\d\.]+', 'match','once'))...
                '-' num2str(regexp(layer_values_names{i}, '[\d\.]+', 'match','once')) ']_km_distribution.png'])
        end

        % Save the matrices based on the binning subdivision 
        [indices,~] = discretize(depth_data,n_bins);
        for j = 1:n_bins
            subdivision_matrix = layer_stress_file(indices == j,:);
            if headers_on == 1
                subdivision_matrix_wheaders = [discretized_headers;num2cell(subdivision_matrix)];
                writecell(subdivision_matrix_wheaders, [subclassified_path '\' matrix_to_save_part ...
                    num2str(regexp(layer_values_names{i}, '[\d\.]+', 'match','once')) ...
                    '_km_bin_' num2str(j) '.csv']);
            else
                writematrix(subdivision_matrix, [subclassified_path '\' matrix_to_save_part ...
                    num2str(regexp(layer_values_names{i}, '[\d\.]+', 'match','once')) ...
                    '_km_bin_' num2str(j) '.csv']);
            end
        end

    end
    
    %----------------------------------------------------------------------

    % We can do the same, which means generating histograms and dividing a
    % larger matrix into binned ones for the entire earth_point part, the
    % one we are interested the most in.
    
    earth_bins = 20;
    [allearth_indices,~] = discretize(earth_data(:,end),earth_bins);
    earth_data_path = [classified_path '\No_layers_subdivision'];
    if ~exist(earth_data_path, 'dir')
        mkdir (earth_data_path)
    end
    if sd_input == 0
        cartesian_coordinates = earth_data(:,9:11);
    else
        cartesian_coordinates = earth_data(:,6:8);
    end
    [lat,lon] = cart2geo(cartesian_coordinates);
    earth_data = [earth_data,lat,lon];
    
    % Save matrices
    
    for j = 1:earth_bins
        subdivision_matrix = earth_data(allearth_indices == j,:);
        if headers_on == 1
            subdivision_matrix_wheaders = [discretized_headers;num2cell(subdivision_matrix)];
            writecell(subdivision_matrix_wheaders, [earth_data_path '\' earth_name_part ...
                'bin_' num2str(j) '.csv']);
        else
            writematrix(subdivision_matrix, [earth_data_path '\' earth_name_part ...
                'bin_' num2str(j) '.csv']);
        end
    end

    cfn = length(findobj('type','figure'));
    % Generating histograms after getting the number of active figures to
    % not overwrite them
    figure(cfn+1)
    histogram(earth_data(:,end-2),earth_bins);
    xlabel('Depth [m]'); ylabel(['Number of ' histogram_name_part]);
    title(['Distribution of ' histogram_name_part ' from the entire ABAQUS model at different depths, n_{bins}='...
        num2str(earth_bins)]);
    grid on;
    saveas(gcf,[histogram_path '\Complete_Earth_distribution.png'])
    
    if headers_on == 1
        earth_matrix_wheaders = [discretized_headers;num2cell(earth_data)];
        writecell(earth_matrix_wheaders, [earth_data_path '\' earth_name_part(1:end-1) '.csv']);
    else
        writematrix(earth_data, [earth_data_path '\' earth_name_part(1:end-1) '.csv']);
    end
end

end

File: python_files_python_3/depth_classifier.py
from natsort import natsorted
from python_discretizer import *
import os
import numpy as np
import re
import matplotlib.pyplot as plt
import csv
import pdb
import cartopy.crs as ccrs


def depth_classifier(sd_input, deflection_processing_path, individual_path, components_to_plot, headers_on,
                     file_extension, cartesian_classified_depth_path, geographical_classified_depth_path,
                     files_to_classify_path, files_to_classify, histogram_path):

    # Here the data is discretized in a certain number of bins, and stored in different matrices for each of those bins.
    # Distinguish two cases, one for the stresses and one for the deflections, and define relevant variables
    if sd_input == 0:
        upper_path = individual_path
        save_name = 'Stresses_until_depth_'
        histogram_name_part = 'centroids'
        matrix_to_save_part = 'Stresses_layer_'
        earth_name_part = 'Stresses_Earth_'
        discretized_headers = ['Label', 'Mises', 'S_11', 'S_22', 'S_33', 'S_12', 'S_13', 'S_23', 'X', 'Y', 'Z', 'R',
                               'Depth', 'Lat', 'Lon']
    else:
        upper_path = deflection_processing_path
        save_name = 'Deflections_until_depth_'
        histogram_name_part = 'points'
        matrix_to_save_part = 'Deflections_layer_'
        earth_name_part = 'Deflections_Earth_'
        discretized_headers = ['Label', 'U_Magn', 'U_1', 'U_2', 'U_3', 'X', 'Y', 'Z', 'R', 'Depth', 'Lat', 'Lon']

    # Extract filenames and create folders based on the components we are working with
    if components_to_plot == 'geographical':
        classified_path = geographical_classified_depth_path
        if not os.path.exists(classified_path):
            os.mkdir(classified_path)
        if not os.path.exists(histogram_path):
            os.mkdir(histogram_path)
    elif components_to_plot == 'cartesian':
        classified_path = cartesian_classified_depth_path
        if not os.path.exists(classified_path):
            os.mkdir(classified_path)
        if not os.path.exists(histogram_path):
            os.mkdir(histogram_path)
    else:
        print('Incorrect input, select either geographical or cartesian.')
    # Only leave EARTH_POINT and EARTH_POINT_LOW as files to bin, because those are the ones that interest us. Do not
    # consider the region files as they have been all saved in a single Earth part file.

    for file_to_evaluate in range(len(files_to_classify)):
        if 'I0' in files_to_classify[file_to_evaluate]:
            files_to_classify[file_to_evaluate] = 'to_delete'
        if 'Large' in files_to_classify[file_to_evaluate]:
            files_to_classify[file_to_evaluate] = 'to_delete'
        if 'Region' in files_to_classify[file_to_evaluate]:
            files_to_classify[file_to_evaluate] = 'to_delete'
        if 'LOW' in files_to_classify[file_to_evaluate]:
            files_to_classify[file_to_evaluate] = 'to_delete'
        if 'coupled' in files_to_classify[file_to_evaluate]:
            files_to_classify[file_to_evaluate] = 'to_delete'
    files_to_classify = [a for a in files_to_classify if a != 'to_delete']
    for file_to_evaluate in range(len(files_to_classify)):
        with open(os.path.join(files_to_classify_path, files_to_classify[file_to_evaluate])) as matrix:
            matrix_to_evaluate = matrix.readlines()
            matrix_to_evaluate = [line.strip() for line in matrix_to_evaluate[1:]]
            matrix_to_evaluate = [np.array([eval(i) for i in line.split(",")[:]]) for line in matrix_to_evaluate]
            matrix_to_evaluate = np.array(matrix_to_evaluate)
            if not matrix_to_evaluate[1:, 1:].any():
                files_to_classify[file_to_evaluate] = 'to_delete'
    files_to_classify = [a for a in files_to_classify if a != 'to_delete']
    maximum_depth = 0
    earth_data = []
    for i in range(len(files_to_classify)):
        with open(os.path.join(files_to_classify_path, files_to_classify[i])) as matrix:
            data_matrix = matrix.readlines()
            data_matrix = [line.strip() for line in data_matrix[1:]]
            data_matrix = [np.array([eval(i) for i in line.split(",")[:]]) for line in data_matrix]
            depth_matrix = np.array(data_matrix)
            maximum_depth = np.amax(depth_matrix[:, -3])
            earth_data.append(depth_matrix)
    earth_data = np.concatenate(earth_data, axis=0)
    if maximum_depth > 2.886e6:
        layer_depth_km = [410, 660, 2886, maximum_depth / 1000]
    else:
        layer_depth_km = [410, 660, maximum_depth / 1000]
    layer_depth_km = np.array(layer_depth_km)
    layer_depth = layer_depth_km * 1000
    check_1 = 0
    check_2 = 1
    run_input = 1
    if os.path.exists(os.path.join(histogram_path, 'Complete_Earth_distribution.png')):
        while check_1 == 1:
            run_input = eval(input('Do you want to rerun the depth discretization algorithm? 1 if yes, 0 if no: \n'))
            if run_input == 0:
                check_1 = 0
            elif run_input == 1:
                while check_2 == 1:
                    bin_input = eval(input('Do you want to change the bin number? 1 if yes, 0 if no: \n'))
                    if bin_input == 0:
                        check_2 = 0
                        check_1 = 0
                    elif bin_input == 1:
                        n_bins = eval(input('Enter number of layer bins:  \n'))
                        earth_bins = eval(input('Enter number of model bins: \n'))
                        check_2 = 0
                        check_1 = 0
                    else:
                        print('Incorrect input, select either 1 or 0. \n')
            else:
                print('incorrect input, select either 1 or 0. \n')
    fig_counter = 1

    if run_input == 0:
        print('Skipping this stage to generate MATLAB handles file.')
    else:
        print('Starting discretization of data into different depths')
        large_depth_matrix = []
        for i in range(len(files_to_classify)):
            with open(os.path.join(files_to_classify_path, files_to_classify[i])) as matrix:
                data_matrix = matrix.readlines()
                data_matrix = [line.strip() for line in data_matrix[1:]]
                data_matrix = [np.array([eval(j) for j in line.split(",")[:]]) for line in data_matrix]
                complete_matrix = np.array(data_matrix)
                large_depth_matrix.append(complete_matrix)
        large_depth_matrix = np.concatenate(large_depth_matrix, axis=0)
        for depth_element in range(len(layer_depth_km)):
            layer_depth_km[depth_element] = float(layer_depth_km[depth_element])

        for i in range(len(layer_depth)):
            layer_group_indices = [j for j, v in enumerate(large_depth_matrix[:, -3]) if v < layer_depth[i]]
            layer_matrix = large_depth_matrix[layer_group_indices, :]
            layer_depth_km_str = str(int(round(layer_depth_km[i])))
            if headers_on == 1:
                with open(os.path.join(classified_path, save_name + layer_depth_km_str + '_km.csv'), 'wb') as \
                        f_write:
                    writer = csv.writer(f_write)
                    writer.writerow(discretized_headers)
                    writer.writerows(layer_matrix)
            else:
                with open(os.path.join(classified_path, save_name + layer_depth_km_str + '_km.csv'), 'wb') as \
                        f_write:
                    writer = csv.writer(f_write)
                    writer.writerows(layer_matrix)
            large_depth_matrix[layer_group_indices, :] = 0
            large_depth_matrix = large_depth_matrix[~np.all(large_depth_matrix == 0, axis=1)]
        n_bins = 30
        layer_values = os.listdir(classified_path)
        layer_values_names = [filename for filename in layer_values if filename.endswith(file_extension)]
        layer_values_names = natsorted(layer_values_names)
        fig_counter = 1
        for i in range(len(layer_values_names)):
            subclassified_path = os.path.join(classified_path, 'depth_' +
                                              str(re.search(r'\d+', layer_values_names[i]).group(0)) + '_km')
            if not os.path.exists(subclassified_path):
                os.mkdir(subclassified_path)
            with open(os.path.join(classified_path, layer_values_names[i])) as layer_matrix_to_discretize:
                layer_data_matrix = layer_matrix_to_discretize.readlines()
                layer_data_matrix = [line.strip() for line in layer_data_matrix[1:]]
                layer_data_matrix = [np.array([eval(j) for j in line.split(",")[:]]) for line in layer_data_matrix]
                layer_data_matrix = np.array(layer_data_matrix)
                depth_data = layer_data_matrix[:, -3]/1000
                plt.figure(fig_counter)
                plt.hist(depth_data, n_bins, edgecolor='k', linewidth=1)
                plt.xlabel('Depth [km]')
                plt.ylabel('Number of points')
                plt.grid(axis='both', alpha=0.75)
                if i == 0:
                    plt.title('Distribution of ' + str(histogram_name_part) +
                              ' from the ABAQUS model at different depths for a depth range of 0 to ' +
                              list(filter(str.isdigit, layer_values_names[i])) + ' km')
                    plt.savefig(os.path.join(histogram_path, 'Depth_range_0_' +
                                             list(filter(str.isdigit, layer_values_names[i]))
                                             + '_km_distribution.png'), bbox_inches='tight')
                    fig_counter += 1
                else:
                    plt.title('Distribution of ' + str(histogram_name_part) +
                              ' from the ABAQUS model at different depths for a depth range of ' +
                              list(filter(str.isdigit, layer_values_names[i - 1])) + ' to ' +
                              list(filter(str.isdigit, layer_values_names[i])) + ' km')
                    plt.savefig(os.path.join(histogram_path, 'Depth_range_' +
                                             list(filter(str.isdigit, layer_values_names[i - 1])) + '-' +
                                             list(filter(str.isdigit, layer_values_names[i])) +
                                             '_km_distribution.png'), bbox_inches='tight')
                    fig_counter += 1
                indices, bin_edges = python_discretizer(depth_data, i, layer_depth, n_bins)
                for j in range(n_bins):
                    subdivision_matrix = layer_data_matrix[indices == j+1, :]
                    smaller_edge = str(int(round(bin_edges[j] / 1000)))
                    if j == n_bins-1:
                        larger_edge = str(int(round(bin_edges[-1] / 1000)))
                    else:
                        larger_edge = str(int(round(bin_edges[j+1] / 1000)))
                    if headers_on == 1:
                        with open(os.path.join(subclassified_path, matrix_to_save_part +
                                        str(re.search(r'\d+', layer_values_names[i]).group(0)) + '_km_bin_' + str(j+1) +
                                        '_edges_' + smaller_edge + '_' + larger_edge + '.csv'), 'wb') as f_write:
                            writer = csv.writer(f_write)
                            writer.writerow(discretized_headers)
                            writer.writerows(subdivision_matrix)
                    else:
                        with open(os.path.join(subclassified_path, matrix_to_save_part +
                                        str(re.search(r'\d+', layer_values_names[i]).group(0)) + '_km_bin_' + str(j+1) +
                                        '_edges_' + smaller_edge + '_' + larger_edge + '.csv'), 'wb') as f_write:
                            writer = csv.writer(f_write)
                            writer.writerows(subdivision_matrix)

    #################################################################################################################

        earth_bins = 50
        data = earth_data[:, -3]
        bins = np.arange(data.min(), data.max()+(data.max()-data.min())/(2*earth_bins), (data.max() - data.min())
                         / earth_bins)
        earth_indices = np.digitize(earth_data[:, -3], bins)
        earth_data_path = os.path.join(classified_path, 'No_layers_subdivision')
        if not os.path.exists(earth_data_path):
            os.mkdir(earth_data_path)

        for j in range(earth_bins):
            subdivision_matrix = earth_data[earth_indices == j+1, :]
            smaller_edge = str(int(round(bins[j] / 1000)))
            if j == n_bins - 1:
                larger_edge = str(int(round(bins[-1] / 1000)))
            else:
                larger_edge = str(int(round(bins[j + 1] / 1000)))
            if headers_on == 1:
                with open(os.path.join(earth_data_path, earth_name_part + 'bin_' + str(j+1) + '_edges_' + smaller_edge +
                                                        '_' + larger_edge + '.csv'), 'wb') as f_write:
                    writer = csv.writer(f_write)
                    writer.writerow(discretized_headers)
                    writer.writerows(subdivision_matrix)
            else:
                with open(os.path.join(earth_data_path, earth_name_part + 'bin_' + str(j+1) + '_edges_' + smaller_edge +
                                                        '_' + larger_edge + '.csv'), 'wb') as f_write:
                    writer = csv.writer(f_write)
                    writer.writerows(subdivision_matrix)

        if headers_on == 1:
            with open(os.path.join(earth_data_path, earth_name_part + '.csv'), 'wb') as f_write:
                writer = csv.writer(f_write)
                writer.writerow(discretized_headers)
                writer.writerows(earth_data)
        else:
            with open(os.path.join(earth_data_path, earth_name_part + '.csv'), 'wb') as f_write:
                writer = csv.writer(f_write)
                writer.writerows(earth_data)

        plt.figure(fig_counter)
        plt.hist(earth_data[:, -3], earth_bins, edgecolor='k', linewidth=1)
        plt.xlabel('Depth [m]')
        plt.ylabel('Number of points')
        plt.title('Distribution of points from the entire ABAQUS model at different depths, n_bins=' + str(earth_bins))
        plt.grid(axis='both', alpha=0.75)
        plt.savefig(os.path.join(histogram_path, 'Complete_Earth_distribution.png'), bbox_inches='tight')
        fig_counter += 1

